ptgnn/baseneuralmodel/distributedtrainer.py

```diff
@@ -0,0 +1,373 @@
+                assert num_samples > 0, "No validation data was found."
+
+            # Sync validation losses
+            validation_loss = sum_epoch_loss / num_minibatches
+            dist.all_reduce(validation_loss)
+            validation_loss = validation_loss.item() / dist.get_world_size()
+
+        except RuntimeError as re:
+            self.LOGGER.exception("Something went wrong: %s", str(re))
+
+        self.LOGGER.info(
+            "Validation complete in %.1fsec [%.2f samples/sec]",
+            elapsed_time,
+            (num_samples / elapsed_time),
+        )
+        self.LOGGER.info("Epoch %i: Valid Loss %.2f", epoch + 1, validation_loss)
+
+        validation_metrics = distributed_neural_module.module.report_metrics()
+        for epoch_hook in self._validation_epoch_end_hooks:
+            epoch_hook(self.model, distributed_neural_module.module, epoch, validation_metrics)
+        if len(validation_metrics) > 0:
+            self.LOGGER.info("Validation Metrics: %s", json.dumps(validation_metrics, indent=2))
+
+        if self._target_metric is not None:
+            target_metric = validation_metrics[self._target_metric]
+            target_metric = torch.tensor([target_metric], device=mb_loss.device)
+            dist.all_reduce(target_metric)
+            target_metric = target_metric.item() / dist.get_world_size()
+        else:
+            target_metric = validation_loss
+
+        if self._target_metric_higher_is_better:
+            target_metric_improved = target_metric > best_target_metric
+        else:
+            target_metric_improved = target_metric < best_target_metric
+        return target_metric, target_metric_improved
+
+    def train(
+        self,
+        training_data: Iterable[TRawDatapoint],
+        validation_data: Iterable[TRawDatapoint],
+        *,
+        show_progress_bar: bool = True,
+        validate_on_start: bool = True,
+        patience: int = 5,
+        initialize_metadata: bool = True,
+        parallelize: bool = True,
+        use_multiprocessing: bool = True,
+        exponential_running_average_factor: float = 0.97,
+        device=None,
+        store_tensorized_data_in_memory: bool = False,
+        shuffle_training_data: bool = True,
+    ) -> None:
+        raise Exception("Use `distributed_train()` instead of calling `train().")
+
+    def distributed_train(
+        self,
+        world_size: int,
+        training_data: ShardedLazyDataIterable[TRawDatapoint],
+        validation_data: ShardedLazyDataIterable[TRawDatapoint],
+        *,
+        validate_on_start: bool = True,
+        patience: int = 5,
+        initialize_metadata: bool = True,
+        parallelize: bool = True,
+        shuffle_training_data: bool = True,
+    ) -> None:
+        """
+        The training-validation loop for `AbstractNeuralModel`s.
+
+        :param training_data: An iterable that each iteration yields the full training data. Note
```

Comment:
I'm not sure I understand this comment. Is the contract that the set of examples by iterating over the iterator on all nodes should be the full training set, or should each node do the full set?

Reply:
(updated)